More systematic binary op tests

SirYwell opened the pull request:
This is more of an experiment to gather some opinions. Please let me know what you think.

With this change, I'm trying to cover interpreter and bytecode generation/execution of binary ops in a more systematic approach.

I didn't find a good way to automatically *generate* test cases, especially as there is insufficient information about allowed types. Therefore, this approach consists of two main steps:

1. A test that ensures that all bin ops have at least one test case
2. Dynamic tests based on these test cases

Note that these tests currently all fail. It seems like neither the interpreter nor the bytecode generator can deal with invokevirtual methods properly, so unboxing always fails.

## Alternatives

One other idea I had was to, instead of going through the dynamic test cases, go through all methods in that class annotated with `@CodeReflection` and check this way if all bin ops are covered. This might allow for more extension to e.g. unary ops. The rest of the tests would then work as in other test classes. The problem of different data types and values would still exist though. Maybe a property-testing approach like `jqwik` would help with different values.

SirYwell commented:
Thanks, that patch resolves the issue. I'll still wait for your opinion on the general structure before adding more test cases.

PaulSandoz commented:
Thanks for confirming (i integrated a PR to fix it). I still need to look more deeply at the testing approach. In the interim here are a few more general comments.

That's an interesting use of `OpDefinition` i never really anticipated, but it does lend itself to more general production of operations. Here is a more concise way. 

```
        CoreOps.LambdaOp lop = ...;
        CoreOps.FuncOp fop = CoreOps.func("test", lop.body().copy(CopyContext.create()));
```

or slightly less concise:

```
        CoreOps.LambdaOp lop = ...;
        CoreOps.FuncOp fop = CoreOps.func("test", lop.invokableType()).body(b -> {
           b.transformBody(lop.body(), b.parameters(), OpTransformer.COPYING_TRANSFORMER);
        });
```


However, we would not have to do this if the bytecode generator more generally accepted an invokable op. We can fix that separately.

SirYwell commented:
> That's an interesting use of `OpDefinition` i never really anticipated, but it does lend itself to more general production of operations. Here is a more concise way (when the lambda does not capture):

Oh, that's what I was looking for, but I somehow missed it. My creation was basically just the first "this works" way I found.

----

I added a second variant, where we use `@CodeReflection`-annotated methods as a base and transform them for different types. This probably makes the distinction between model and test input clearer. But for now, I'm not 100% happy with either solution.

----

Btw, I noticed that `FuncOp#resultType()` always `JavaType.VOID`. Is that intended? (At least I was surprised)

PaulSandoz commented:
> I added a second variant, where we use `@CodeReflection`-annotated methods as a base and transform them for different types. This probably makes the distinction between model and test input clearer. But for now, I'm not 100% happy with either solution.

I feel the same way, preferring second variant because of the distinction. I also like that it is leveraging a transformation, replacing the types. This pushes on an area i don't think we have fully worked out. Not all operations derive their result type from their operands types. In general transforming to replace types is i think challenging but can work well in these cases.

As an experiment it seems worth perusing further even if it is not generally applicable as i think we might learn things on the way. We could lean into the second variant, and you can probably drop `ensureAllBinOpsCovered`.

There might be a way in JUnit to annotate the code reflected method as a test template and then annotate the test class with `@ExtendWith`, such that we extend JUnit to execute the code reflected method as we wish similarly to the dynamic approach, but without needing to reflectively enumerate all the methods under test. 

  https://junit.org/junit5/docs/current/user-guide/#extensions
  https://junit.org/junit5/docs/current/api/org.junit.jupiter.api/org/junit/jupiter/api/extension/TestTemplateInvocationContextProvider.html

PaulSandoz commented:
> Btw, I noticed that `FuncOp#resultType()` always `JavaType.VOID`. Is that intended? (At least I was surprised)

Yes, my thinking was that function declaration operations don't return something that can be interacted with directly and cannot be passed around as values. We can however produce a symbol table of function name to function (e.g., for functions that are grouped into a module) and there is an operation to call a function by name (which will have stack effects). Functions are not intended to be part of some block of code and they do not capture values (we refer to the body as isolated). They are intended to model Java methods. Lambda and closure operations can be embedded in code and both can capture, the former returns an instance of a functional interface, and the latter returns an instance of a function type (which has no direct correlation to Java source).

SirYwell commented:
Thanks for your comments.

> There might be a way in JUnit to annotate the code reflected method as a test template and then annotate the test class with `@ExtendWith`, such that we extend JUnit to execute the code reflected method as we wish similarly to the dynamic approach, but without needing to reflectively enumerate all the methods under test.

I explored a bit how we can use JUnit extensions. One downside of directly annotating a `@CodeReflection` method as a test is that such methods need to be non-static (at least from what I've found out), which makes transforming more complicated. Alternatively, we could have one method that is a `@ParameterizedTest` with an argument provider that deals with the transformation and also provides the test input. I don't have something working for now, but that could look like:
```java
    @ParameterizedTest
    @CodeReflectionSource
    void test(CoreOps.FuncOp funcOp, TestInput testInput, Result expectedResult) {
        Result interpret = runCatching(() -> interpret(testInput, funcOp));
        Result bytecode = runCatching(() -> bytecode(testInput, funcOp));
        assertResults(expectedResult, interpret);
        assertResults(interpret, bytecode);
    }
```
I think such a design could also be used for e.g. unary ops. The downside is that we still need to look for `@CodeReflection` methods ourselves. The upside is that it's clearer what the test is actually doing.

What's your opinion on that?

----

> Yes, my thinking was that function declaration operations don't return something that can be interacted with directly and cannot be passed around as values. We can however produce a symbol table of function name to function (e.g., for functions that are grouped into a module) and there is an operation to call a function by name (which will have stack effects). Functions are not intended to be part of some block of code and they do not capture values (we refer to the body as isolated). They are intended to model Java methods. Lambda and closure operations can be embedded in code and both can capture, the former returns an instance of a functional interface, and the latter returns an instance of a function type (which has no direct correlation to Java source).

That thinking absolutely makes sense. Maybe it's enough to add some docs, but currently I would expect others to also be confused by the difference between `resultType()` and `invokableType().returnType()`. Otherwise, splitting up `Op` into more interfaces would probably make sense, but that's something that most likely requires more exploration of the overall concept I guess.

PaulSandoz commented:
> What's your opinion on that?

If we lean into a test template approach with transformations then dealing with a template that is an instance method does not seem too difficult, just drop the first block parameter when transforming (any uses of that parameter means a test infrastructure bug) or pass the test instance argument.

The parameterized approach would i think be more familiar. If you like and have the time perhaps experiment with both appraoches? Perhaps there are valid uses for both kinds of approaches? Otherwise pick the one you think is better. I expect we might iterate on this area in subsequent PRs.

PaulSandoz commented:
> 
> That thinking absolutely makes sense. Maybe it's enough to add some docs, but currently I would expect others to also be confused by the difference between `resultType()` and `invokableType().returnType()`. Otherwise, splitting up `Op` into more interfaces would probably make sense, but that's something that most likely requires more exploration of the overall concept I guess.

I wonder if, understandably, when initially encountering operation it is thought about in the more conventional sense e.g., add two numbers? But, it does not have to be. We can define an operation that *declares* a function, and its result type is certainly not the type of what the function returns when invoked, and it may not have a meaningful result (hence `void`, which i admit adds to the confusion, and perhaps `FunctionType` is more honest which would be equivalent to that returned by `invokableType`).

We do have a common interface that abstracts invokable operations, `Op.Invokable`, and that declares method `invokableType`.

SirYwell commented:
> If we lean into a test template approach with transformations then dealing with a template that is an instance method does not seem too difficult, just drop the first block parameter when transforming (any uses of that parameter means a test infrastructure bug) or pass the test instance argument.
> 
> The parameterized approach would i think be more familiar. If you like and have the time perhaps experiment with both appraoches? Perhaps there are valid uses for both kinds of approaches? Otherwise pick the one you think is better. I expect we might iterate on this area in subsequent PRs.

I'll try out both to see how it turns out, but it might take a few days.

SirYwell commented:
> I wonder if, understandably, when initially encountering operation it is thought about in the more conventional sense e.g., add two numbers? 

As `FuncOp` is one of the two main entry points to the API (at least from my view, besides going through `Quoted`), I think that's not too much of a concern overall.

> But, it does not have to be. We can define an operation that _declares_ a function, and its result type is certainly not the type of what the function returns when invoked, and it may not have a meaningful result (hence `void`, which i admit adds to the confusion, and perhaps `FunctionType` is more honest which would be equivalent to that returned by `invokableType`).
> 
> We do have a common interface that abstracts invokable operations, `Op.Invokable`, and that declares method `invokableType`.

When comparing to `LambdaOp`, the meaning of `resultType` is clearer as we just have an expression there. I think the main source of my confusion was that `resultType` showed up directly in the IDE while getting to the return type is more difficult, combined with the fact that methods in Java are neither expressions nor statements, so it's kind of "what else would it be if not the return type?" But I also see how it makes sense to have it modeled this way, it still has most things in common with other Ops and a more fine-grained type hierarchy might be even more confusing or annoying to work with.

I wonder if there are more examples of Ops that have a similar special position. I assume something like local classes/interfaces won't be modeled at all, and e.g. `assert` also returns void, where it makes sense I guess. (Or maybe it could be reasonable to have a separate Unit type? That might be more obvious when used wrongly, e.g. a `void` method currently returns `void` for `resultType`...)

That's just a few thoughts, maybe it's something we shouldn't be too concerned with for now.

PaulSandoz commented:
> ... the fact that methods in Java are neither expressions nor statements

Right.
 
> I wonder if there are more examples of Ops that have a similar special position. I assume something like local classes/interfaces won't be modeled at all, and e.g. `assert` also returns void, where it makes sense I guess. (Or maybe it could be reasonable to have a separate Unit type? That might be more obvious when used wrongly, e.g. a `void` method currently returns `void` for `resultType`...)
>

Yeah, i think a code model unit type is clearer (rather than reusing `JavaType.VOID`, which is a likely source of confusion).

There's a "module" operation which has one body with one block whose operations should all be function declaration operations. This is similar to what MLIR does. That concept is used in the Triton example (see [here](https://github.com/openjdk/babylon/blob/code-reflection/cr-examples/triton/src/test/java/oracle/code/triton/TestSoftMax.java#L39)). But, we have yet to use it for modeling Java code.

We don't intend to model class declarations -- i would very much like to avoid that. We are still working out exactly what to do when reflecting on code that has class declarations. It may be we surface up all the associated reflected methods as function declaration operations in a module operation.

Another place where module operations could appear is if someone wrote a transformation to translate lambdas similarly to what java compiler does when generating bytecode (the lambda is desugared into an invokedynamic a referenced method).

SirYwell commented:
I pushed a commit containing the approach using `@ParameterizedTest` in `CoreBinaryOpsTest`. I also included a bit more useful input values than before. I also adjusted some things to theoretically make it work with unary ops and test ops. The current implementation would also allow to use more specific types in the parameters of a parameterized test. I don't know if this is needed, but I could imagine that this could be helpful when e.g. wanting to make sure that floating point operations result in the same raw bits (e.g. using `Double#doubleToRawLongBits`).

Regarding the `@TestTemplate` approach, such methods don't only need to be non-static but also `void`, and the intention is to call exactly that method one or multiple times. So I think that's not what we want. `@Testable` is less restrictive, but from my understanding isn't picked up by the default junit test engine. Unless I missed something there, I think it's not worth further going into that direction.

PaulSandoz commented on test/jdk/java/lang/reflect/code/CoreBinaryOpsTest.java:
```diff
@@ -0,0 +1,270 @@
+            if (first.throwable.getClass() != second.throwable.getClass()) {
+                first.throwable.printStackTrace();
+                second.throwable.printStackTrace();
+                fail("Different exceptions where thrown");
```
```suggestion
                fail("Different exceptions were thrown");
```

PaulSandoz commented on test/jdk/java/lang/reflect/code/CoreBinaryOpsTest.java:
```diff
@@ -0,0 +1,270 @@
+                second.throwable.printStackTrace();
+                fail("Different exceptions where thrown");
+            }
+            assertEquals(first.throwable.getClass(), second.throwable.getClass());
```
Never fails? Due to the prior equality check.

PaulSandoz reviewed (commented):
`CoreBinaryOpsTest` looks good. Recommend you place the reflected template methods and the test method as close to the top of the class as possible.

SirYwell commented:
I moved the relevant methods to the top. All the code surrounding the JUnit extension could be moved into separate files, allowing to use it in other test classes as well (but I'd need to figure out how that works with jtreg and find a good place for those files).

I also added code for other implemented binary ops now, as the test design is more stable now I guess.

PaulSandoz commented:
Very good. We can separate out later if need be (should be simple if source resides in the same package) and when we see an opportunity to adapt and reuse it (we might when a PR is published for string concatenation).

SirYwell commented:
> Very good. We can separate out later if need be (should be simple if source resides in the same package) and when we see an opportunity to adapt and reuse it (we might when a PR is published for string concatenation).

Okay, thanks. I would continue with implementing more missing binary ops (namely shift operations) unless you have any specific plans for that.

/integrate
